### PR TITLE
[13.0][IMP] sale_by_packaging

### PR DESCRIPTION
**Pull request by TDu**
Changes detailed in commit messages

### CODE REVIEW

File: sale_by_packaging/models/sale_order_line.py [hunk 1]
```diff
@@ -39,14 +39,16 @@ def _onchange_product_packaging(self):
...
         return super()._onchange_product_packaging()
 
-    @api.constrains("product_id", "product_packaging")
+    @api.constrains("product_id", "product_packaging", "product_packaging_qty")
     def _check_product_packaging_sell_only_by_packaging(self):
         for line in self:
-            if line.product_id.sell_only_by_packaging and not line.product_packaging:
+            if not line.product_id.sell_only_by_packaging:
+                continue
+            if not line.product_packaging or line.product_packaging_qty <= 0:
```
Comment (rousseldenis) [inline sale_by_packaging/models/sale_order_line.py]: As product_packaging_qty is managed with digit precision, you should use float_compare()

File: sale_by_packaging/models/sale_order_line.py [hunk 2]
```diff
@@ -88,6 +90,9 @@ def _onchange_product_uom_qty(self):
         res = super()._onchange_product_uom_qty()
         if not res:
             res = self._check_qty_is_pack_multiple()
+        if "warning" in res.keys():
```
Comment (rousseldenis) [inline sale_by_packaging/models/sale_order_line.py]: Is it really a good test having "warning" to reset those fields ?

**Review by rousseldenis — CHANGES_REQUESTED** [review]
LGTM. At least the float_compare() change

File: sale_by_packaging/models/sale_order_line.py (hunk 1 above)
Comment (TDu) [reply sale_by_packaging/models/sale_order_line.py]: Thanks for your review will change that

File: sale_by_packaging/models/sale_order_line.py (hunk 2 above)
Comment (TDu) [reply sale_by_packaging/models/sale_order_line.py]: Nope, definitely needs to be improved

Comment (TDu) [reply sale_by_packaging/models/sale_order_line.py]: I changed it, also the test is still on the return value (warning) now it is only based on one function. Much clearer and safer.

File: sale_by_packaging/models/sale_order_line.py (hunk 1 above)
Comment (TDu) [reply sale_by_packaging/models/sale_order_line.py]: Changed

**Review by guewen — APPROVED** [review]

**Review by sebalix — APPROVED** [review]

**Comment by guewen** [comment]
@TDu can you squash your fixup?

If we merge this one, we'll be able to rebase https://github.com/OCA/sale-workflow/pull/1276 to resolve conflicts.

**Comment by guewen** [comment]
@rousseldenis could you update your review?

File: sale_by_packaging/models/sale_order_line.py [hunk 3]
```diff
@@ -86,9 +96,11 @@ def _force_qty_with_package(self):
...
-        if not res:
-            res = self._check_qty_is_pack_multiple()
-        return res
+        is_pack_multiple_warning = self._check_qty_is_pack_multiple()
+        if is_pack_multiple_warning:
+            self.product_packaging_qty = False
+            self.product_packaging = False
+        return res if res else is_pack_multiple_warning
 
     def _check_qty_is_pack_multiple(self):
```
Comment (simahawk) [inline sale_by_packaging/models/sale_order_line.py:105]: IMO this _check_qty_is_pack_multiple is totally useless as the validation is already done by odoo core.
I removed it here https://github.com/OCA/sale-workflow/pull/1231/ :thinking:

**Pull request merged**